src/texture/any.rs

```diff
@@ -69,6 +69,53 @@ pub struct TextureAny {
     levels: u32,
     /// Is automatic mipmap generation allowed for this texture?
     generate_mipmaps: bool,
+
+    /// Is this texture owned by us? If not, we won't clean it up on drop.
+    owned: bool
+}
+
+fn extract_dimensions(ty: Dimensions)
+                      -> (u32, Option<u32>, Option<u32>, Option<u32>, Option<u32>)
+{
+    match ty {
+        Dimensions::Texture1d { width } => (width, None, None, None, None),
+        Dimensions::Texture1dArray { width, array_size } => (width, None, None, Some(array_size), None),
+        Dimensions::Texture2d { width, height } => (width, Some(height), None, None, None),
+        Dimensions::Texture2dArray { width, height, array_size } => (width, Some(height), None, Some(array_size), None),
+        Dimensions::Texture2dMultisample { width, height, samples } => (width, Some(height), None, None, Some(samples)),
+        Dimensions::Texture2dMultisampleArray { width, height, array_size, samples } => (width, Some(height), None, Some(array_size), Some(samples)),
+        Dimensions::Texture3d { width, height, depth } => (width, Some(height), Some(depth), None, None),
+        Dimensions::Cubemap { dimension } => (dimension, Some(dimension), None, None, None),
+        Dimensions::CubemapArray { dimension, array_size } => (dimension, Some(dimension), None, Some(array_size * 6), None),
+    }
+}
+
+#[inline]
+fn get_bind_point(ty: Dimensions) -> gl::types::GLenum {
+    match ty {
+        Dimensions::Texture1d { .. } => gl::TEXTURE_1D,
+        Dimensions::Texture1dArray { .. } => gl::TEXTURE_1D_ARRAY,
+        Dimensions::Texture2d { .. } => gl::TEXTURE_2D,
+        Dimensions::Texture2dArray { .. } => gl::TEXTURE_2D_ARRAY,
+        Dimensions::Texture2dMultisample { .. } => gl::TEXTURE_2D_MULTISAMPLE,
+        Dimensions::Texture2dMultisampleArray { .. } => gl::TEXTURE_2D_MULTISAMPLE_ARRAY,
+        Dimensions::Texture3d { .. } => gl::TEXTURE_3D,
+        Dimensions::Cubemap { .. } => gl::TEXTURE_CUBE_MAP,
+        Dimensions::CubemapArray { .. } => gl::TEXTURE_CUBE_MAP_ARRAY,
+    }
+}
+
+unsafe fn generate_mipmaps(ctxt: &CommandContext,
```

Comment:
Open question: where mipmaps are regenerated on upload, this logic is slightly different (the GLES check is not in place). Is that a bug, or intentional?